3634/0/about_sandwich_code.rb

```diff
@@ -0,0 +1,102 @@
+# rubocop:disable Security/Open, Lint/AssignmentInCondition
+# rubocop:disable Performance/RedundantMatch, Lint/UnneededCopDisableDirective
+require File.expand_path(File.dirname(__FILE__) + '/neo')
+# :reek:NilCheck
+class AboutSandwichCode < Neo::Koan
+  def count_lines(file_name)
+    file = open(file_name)
+    count = 0
+    count += 1 while file.gets
+    count
+  ensure
+    file&.close
+  end
+
+  def test_counting_lines
+    assert_equal 4, count_lines('example_file.txt')
+  end
+
+  # ------------------------------------------------------------------
+  # :reek:NilCheck
+  def find_line(file_name)
+    file = open(file_name)
+    while line = file.gets
+      return line if line =~ /e/
+    end
+  ensure
+    file&.close
+  end
+
+  def test_finding_lines
+    assert_equal "test\n", find_line('example_file.txt')
+  end
+
+  # ------------------------------------------------------------------
+  # THINK ABOUT IT:
+  #
+  # The count_lines and find_line are similar, and yet different.
+  # They both follow the pattern of "sandwich code".
+  #
+  # Sandwich code is code that comes in three parts: (1) the top slice
+  # of bread, (2) the meat, and (3) the bottom slice of bread.  The
+  # bread part of the sandwich almost always goes together, but
+  # the meat part changes all the time.
+  #
+  # Because the changing part of the sandwich code is in the middle,
+  # abstracting the top and bottom bread slices to a library can be
+  # difficult in many languages.
+  #
+  # (Aside for C++ programmers: The idiom of capturing allocated
+  # pointers in a smart pointer constructor is an attempt to deal with
+  # the problem of sandwich code for resource allocation.)
+  #
+  # Consider the following code:
+  #
+  # :reek:NilCheck
+  def file_sandwich(file_name)
+    file = open(file_name)
+    yield(file)
+  ensure
+    file&.close
+  end
+
+  # Now we write:
+
+  def count_lines_two(file_name)
+    file_sandwich(file_name) do |file|
+      count = 0
```

Comment:
this `sandwich` could be changed to:
```
file.each_line.inject(0) { |size| size += 1}
```
or
```
file.readlines.size
```
answer in DM about difference between this two options

Reply:
In the first case the method `each_line` splits string from the file  into paragraphs delimited by multiple successive newlines. After that the `inject` method accumulates a result that turns out over the block the `inject` method takes.

In the second case the `readlines` method count all of the lines from the file and returns an array. The `size` method counts items amount of the array and returns this amount.